src/runtime/strategies/match-recipe-by-verb.ts

```diff
@@ -232,24 +240,26 @@ export class MatchRecipeByVerb extends Strategy {
 
   static satisfiesSlotConnection(recipe, slotName, constraints) {
     for (const particle of recipe.particles) {
+      if (!particle.spec) continue;
       if (MatchRecipeByVerb.slotsMatchConstraint(
-              particle.consumedSlotConnections, slotName, constraints)) {
+              particle, particle.spec.slots, slotName, constraints)) {
         return true;
       }
     }
     return false;
   }
 
-  static slotsMatchConstraint(connections, name, constraints) {
-    if (connections[name] == null) {
+  static slotsMatchConstraint(particle, slotSpecs, name, constraints) {
+    if (slotSpecs.get(name) == null) {
       return false;
     }
-    if (connections[name]._targetSlot != null &&
+    const slotConn = particle.getSlotConnectionBySpec(slotSpecs.get(name));
+    if (slotConn && slotConn._targetSlot != null &&
         constraints.targetSlot != null) {
       return false;
     }
     for (const provideName in constraints.providedSlots) {
-      if (connections[name].providedSlots[provideName] == null) {
+      if (slotSpecs.get(name).providedSlots.find(provideSlotSpec => provideSlotSpec.name === provideName).length === 0) {
```

Comment:
General comment: it is often useful to use terse names for variables used in such a self-contained context, to aid readability: `.providedSlots.find(spec => spec.name === provideName)`

Reply:
Done :)